design/ceph/osd-migration.md

```diff
@@ -0,0 +1,106 @@
+# OSD MIGRATION
+
+## SUMMARY
+- Allow new OSDs to use a new storage backend called bluestore_rdr or seastore.
+- Migrate existing OSDs to use a new backend called bluestore_rdr or seastore. The migration process should include destroying the existing OSDs one by one, wiping the disk and then creating a new OSD on that disk.
+
+
+## MOTIVATION:
+- Ceph uses BlueStore (since Luminous) as a special-purpose storage back end designed specifically for managing data on disk for Ceph OSD workloads. Ceph team is working on improving this backend storage that will help in significantly improving the performance, especially in disaster recovery scenarios. This new OSD backend is known as bluestore_rdr or seastore.
+- To support any new and improved storage backend, the existing OSD needs to be destroyed, disks should be wiped out and new OSD should be created using the latest storage backend.
+
+## GOALS
+- Newly provisioned clusters should be able to use the specific storage backend while creating OSDs.
+- Existing clusters should be able to upgrade the OSDs to use a specific storage backend, with a minimal downtime and without losing their data.
+
+
+## PROPOSAL
+- Add a new field called `spec.Storage.osdStore` in the ceph cluster yaml.
+- This will contain:
+  - `spec.Storage.osdStore.type`: The type of backend store to be used for OSDs. For example, bluestore, bluestore_rdr, etc. The default store type will be `bluestore`
+  - `spec.Storage.osdStore.updateStore`: Whether the user wants to migrate the existing OSDs to use new storage backend. The migration process will involve destroying the exsisting OSDs one by one, wiping the disks, creating a new OSD and waiting for the PGs to be active+clean before migrating the next OSD. Since this operation involves possible downtime, users should be really sure before proceeding with this action. This field can only take the value as `yes-really-update-store`
+  -
+
+- Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore_rdr
+    updateStore: yes-really-update-store
```

Comment:
what will be impact of supporting changing to the previous backend store, say from `bluestore_rdr` to `bluestore`?
Since we are creating new OSDs, I can't see why should not support this. Since the bluestore_rdr is new and might have issues, the user might back to switch back to the previous store.  What do you think?

Reply:
ceph won't support this.